Fix #20338: Replace $timeout in exploration metadata modal

ojasvi004 opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #20338.
2. This PR does the following: This PR removes the use of $timeout in the `ExplorationMetadataModalComponent` and replaces it with `Promise.all()`. This change ensures everything runs smoothly without timing issues.
3. (For bug-fixing PRs only) The original bug occurred because: The bug occurred because `$timeout` was being used, which could lead to potential race conditions. This made the modal's behavior less predictable when saving changes.
## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [ ] I have written tests for my code.
- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).


## Proof that changes are correct

[Screencast from 2025-02-03 23-42-13.webm](https://github.com/user-attachments/assets/e129d89c-316c-481f-8e60-fbb7070a5236)

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

ojasvi004 committed:
remove timeout from exploration metadata modal

Vir-8 commented on core/templates/pages/exploration-editor-page/modal-templates/exploration-metadata-modal.component.ts:
```diff
@@ -149,38 +148,43 @@ export class ExplorationMetadataModalComponent
 
     // Record any fields that have changed.
     let metadataList: string[] = [];
+    const savePromises: Promise<void>[] = [];
```
`savePromises` is an extremely vague name + array names shouldn't start with verbs such as `save`, can we give it a better name?

ojasvi004 replied:
I've made some changes in the approach.

Vir-8 reviewed (changes requested):
Hi @ojasvi004, thanks for the PR! One minor comment.

ojasvi004 committed:
update array name

ojasvi004 committed:
update: remove timeout from exploration metadata modal

ojasvi004 committed:
Merge branch 'oppia:develop' into fix-timeout-metadata-modal-20338

ojasvi004 committed:
update

ojasvi004 committed:
update save method to use async/await for sequential metadata saving

ojasvi004 committed:
Merge branch 'oppia:develop' into fix-timeout-metadata-modal-20338

ojasvi004 committed:
Merge branch 'oppia:develop' into fix-timeout-metadata-modal-20338

ojasvi004 committed:
remove setTimeout

ojasvi004 commented:
@Vir-8, I've updated the code and removed the `setTimeout()`
The function `saveDisplayedValue()` runs synchronously. Even though the eventEmitter is asynchronous, it triggers an event that is handled immediately within the same flow (before `saveDisplayedValue` completes). This predictable order eliminates race conditions, making `setTimeout()` unnecessary.

ojasvi004 commented:
@Vir-8, I've addressed the previous requested changes. Please review this PR. Thanks!

Vir-8 commented on core/templates/pages/exploration-editor-page/modal-templates/exploration-metadata-modal.component.ts:
```diff
@@ -172,15 +172,7 @@ export class ExplorationMetadataModalComponent
-    setTimeout(() => {
-      this.ngbActiveModal.close(metadataList);
-    }, 500);
+    this.ngbActiveModal.close(metadataList);
```
Ummm, wouldn't removing the timeout reintroduce the previous race condition and cause issues? Please elaborate.

The PR description seems very inaccurate now, I don't see your mentioned approach being used.

Vir-8 reviewed (changes requested):
Hi @ojasvi004, in the future, please assign the relevant reviewers to the PR instead (in the assignees tab) because it's difficult to keep track of PRs otherwise. Thanks!
